reference/concepts/functions.md

```diff
@@ -1,6 +1,8 @@
 ## Functions
 
-By "functions" we mean functions, methods and procedures.
+By "functions" we mean functions, procedures and sub-routines.
```

Comment:
Where did 'methods' go?  Can we bring that back in, as it addresses some languages.

Reply:
@kotp point of the PR being that methods has it's own concept file. That concept file links to this one. The resulting logical concept explanation being  "A method is a method".
Using "method" to define "method" is the circularity I am trying to remove.

Reply:
See: https://github.com/exercism/v3/blob/master/reference/concepts/methods.mdl

Reply:
Yeah, I get that, but we are circularly referencing function in the same statement here...

Reply:
Shall we just get rid of this sentence and split it into three files? (Functions, methods and subroutine)

Reply:
Perhaps, it seems to be that they are pretty specific things in languages, and not always available to all languages.

Reply:
Apologies @kotp, I misunderstood your meaning the first time around 😳 

Are their languages that have more than one of {functions, subroutines, procedures} and that differentiates between them? I am not a programming polyglot but I can say that, for example, in assembly functions, methods and procedures mostly don't exist, but does have routine/subroutine which is the language specific equivalent of them.

Perhaps this concept should either be merged with the method one, or it can be renamed to be about functional statement grouping more generally. Then the method, function, subroutine, etc concepts can all, separately, link to that?

Reply:
Yes, there are languages that have more than one of function, method, procedure, routine, sub-routine, etc.  And there are some languages that don't have one or more of those things.  There are languages that refer to some of these things in different ways as well. 

Now, "functional statement grouping" is perhaps a totally different subject... and until you mentioned it, it is not something that I had in my mind regarding this conversation.  But it is something I would think of if I were working on something in Assembly, for example.  When thinking in terms of another language, perhaps "behavior grouping" would be how I might think of that in a similar way.

Reply:
Polyglot here! TL;DR, because the overlap but also difference between all these, some papers use the term `callable unit` to refer to _any_ of these. The implementations differ greatly. e.g. sometimes functions, methods, lambda and procedures are all subroutines, but sometimes a subroutine is distinctly different than a function in terms of return value.

> Are their languages that have more than one of {functions, subroutines, procedures} and that differentiates between them? I am not a programming polyglot but I can say that, for example, in assembly functions, methods and procedures mostly don't exist, but does have routine/subroutine which is the language specific equivalent of them.

- SQL has Functions which _must_ return a value. They are _computed values_ and should be side-effect free. They're not allowed to change the environment (so no inserts/deletes/updates).
- SQL has Procedures which allow for 0 to many return values, usually have side-effects (that's what they're for).

- .NET (VB) has Functions. They can return a value
- .NET has Subroutines. They can not return a value

---

They are all valid names, and should all have files, and I don't think cycles are a problem. 

Then there are selections of at least 0 of the following:

- co-routines (sub-routines that can yield/have multiple entry points)
- closures (in all shapes and forms, such as procs, lambdas, anonymous functions)
- blocks (syntax that allows to yield and/or return, grouping statements)
- generic functions (different from there non-generic counterpart)
- overloaded functions (yes, support and usage is different across languages!, but also different _ways_ of function overloading: dynamic dispatch)
- operator functions / operator overloading

Reply:
> They are all valid names, and should all have files, and I don't think cycles are a problem.

I agree with this. I don't mind the cycles, especially not if we correctly link to the various documents.